packages/core/state/selection/logics.ts

```diff
@@ -340,6 +344,53 @@ const toggleFileFolderCollapse = createLogic({
     type: [TOGGLE_FILE_FOLDER_COLLAPSE],
 });
 
+/**
+ * Interceptor responsible for transforming COLLAPSE_ALL_FILE_FOLDERS and EXPAND_ALL_FILE_FOLDERS
+ * actions into SET_OPEN_FILE_FOLDERS actions by either setting to none or recursively
+ * unpacking the directory structure
+ */
+const toggleAllFileFolders = createLogic({
+    async process(deps: ReduxLogicDeps, dispatch, done) {
+        const { action, getState } = deps;
+        const hierarchy = selection.selectors.getAnnotationHierarchy(getState());
+        const annotationService = interaction.selectors.getAnnotationService(getState());
+        const selectedFileFilters = selection.selectors.getFileFilters(getState());
+        const fileFoldersToOpen: FileFolder[] = [];
+        dispatch(setIsLoadingFullTree(true) as AnyAction);
+        // Recursive helper
+        async function unpackAllFileFolders(values: string[], pathSoFar: string[]) {
+            for (const value of values) {
+                fileFoldersToOpen.push(new FileFolder([...pathSoFar, value] as AnnotationValue[]));
+
+                // At end of folder hierarchy
+                if (!!hierarchy.length && pathSoFar.length === hierarchy.length - 1) continue;
+
+                const childNodes = await annotationService.fetchHierarchyValuesUnderPath(
+                    hierarchy,
+                    [...pathSoFar, value],
+                    selectedFileFilters
+                );
+                if (childNodes.length) {
+                    // Not a leaf
+                    unpackAllFileFolders(childNodes, [...pathSoFar, value]);
```

Comment:
What is the performance of this looking like? I wonder if this should at every level dispatch a `setOpenFileFolders` action to save the folders it has found so far. However, that would interfere with the `isLoadingFullTree` flag you have (which I asked about in another comment).

Reply:
My tests were generally taking less than 30 seconds, but I can see it taking longer. Will try this

Reply:
version of this done in [c16c1ae](https://github.com/AllenInstitute/biofile-finder/pull/446/commits/c16c1ae522161412502360a675bab15dbacbf44f)